### DISCUSSION TITLE
Box Annotation coordinates

### DISCUSSION THREAD

**Discussion by afshinmoshrefi** [chartjs/chartjs-plugin-annotation#775] [General]
I have a box annotation drawn on a line chart.

I have an event handler that needs to know when the cursor is inside the annotation box.

To accomplish this, I need to calculate the screen coordinates of the box annotation.  

How can I calculate the screen location of the box annotation?

**Comment by stockiNail** [comment]
@afshinmoshrefi all annotations can have some events hooks (see doc https://www.chartjs.org/chartjs-plugin-annotation/latest/guide/configuration.html#events).

In your use case, it could be enough to set the `enter` hook:

```javascript
plugins: {
  annotation: {
    annotations: {
      myBox: {
        type: 'box',
        ...
        enter(context, event) {
          // cursor hover the annotation
        }
      }
    } 
  }
}
```

**Reply by afshinmoshrefi** [reply]
Thanks, This is just what I was looking for.  I looked but couldn't find it.

**Reply by afshinmoshrefi** [reply]
I have it working by capturing the annotation box coordinates after enter event is triggered.
How can I get access to the annotation box coordinates before enter event is triggered?

**Reply by stockiNail** [reply]
At the moment there isn't any public API which enables you to access to the element of an annotation.
Is the same use case you reported at the beginning of the discussion?

**Reply by afshinmoshrefi** [reply]
yes - I have a react app displaying a chart with annotation.  I'm creating an interactive feature where users can change the size of the annotation box.  All is working, the only issue is I get an unnecessary render when the mouse pointer goes over the annotation box because of a state change.  If I could access the annotation element earlier, I could avoid the extra render.

**Reply by stockiNail** [reply]
> the only issue is I get an extra render when the mouse pointer goes over the annotation box because of a state change. 

This is what it's weird to me. The plugin uses `beforeEvent` hook of Chart.js to catch and manage events. And it doesn't do any rendering. Let me prepare a codepen with your use case, starting from https://www.chartjs.org/chartjs-plugin-annotation/master/samples/interaction/dragging.html

**Reply by stockiNail** [reply]
Codepen: https://codepen.io/stockinail/pen/gOeGOmK
It's not completed but pressing the mouse pointer on box annotation, and move the mouse, the size of annotation is changing.

**Reply by afshinmoshrefi** [reply]
The extra render issue is due to the way React works.  When there is a state variable change, there will be a render even if it's unnecessary. 
 I'm setting the coordinates of the annotation box in a React state variable when the "enter" event on the annotation box is triggered.
That's why I was hoping to get access to the annotation element ahead of time so that I didn't have to update state variables after the "enter" event on the annotation is triggered.  So I'm getting an extra render because I can only get the coordinates after the first "enter" event is triggered.

**Reply by stockiNail** [reply]
OK, clear. Unfortunately there isn't any public API to get the elements.

**Reply by stockiNail** [reply]
Probably a possible solution is to wait for version 2.1.0 where PR #744 could be added. Activating a hook, you can get the elements before or after drawing.

**Reply by afshinmoshrefi** [reply]
When do you think version 2.1.0 will be released?

**Reply by stockiNail** [reply]
Version 2.0.0 has been released 8 days ago therefore some weeks are needed I guess.

Anyway, have a look to the test suite (I think it can help you even if, be aware, it isn't documented and maintained for public use): https://github.com/chartjs/chartjs-plugin-annotation/blob/1857f42c4d0db556fb8ebc8e4672b3a45fa77e96/test/utils.js#L18-L20
